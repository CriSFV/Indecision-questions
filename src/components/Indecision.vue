<template>
<div>
     <img v-show="img" :src="img">
  <div class="bg-dark"></div>
  <div class="indecision-container">
      <input type="text" placeholder="Realiza una pregunta" v-model="question">
      <p>Recuerda terminar con un signo de interrogación (?)</p>
      <div v-if="isValidQuestion">
          <h2>{{ question }}</h2>
          <!-- <h1 v-if="answer==='yes'"> Si </h1>
          <h1 v-else> {{answer}} </h1> -->
          <h1 > {{answer}} </h1>
      </div>
  </div>
</div>
 
</template>

<script>
export default {
    name: 'Indecision',
    data(){
        return {
            question:'',
            answer: null,
            img: '',
            isValidQuestion:false
        }
    },
    methods:{
        async getAnswer(){

            try{
                this.answer='Pensando...'
                const lang = {
                    yes: 'Si!',
                    no: 'No!',
                    maybe: 'Quizás...'
                }
                const {answer, image} = await fetch('https://yesno.wtf/api').then(response=> response.json())
       
                this.answer = lang[answer] 
                this.img = image
            }  catch(error){
                console.log('IndecisionConsoleComponent Error', error)
                this.answer = 'No se pudo cargar la respuesta del API'
                this.img = null
            }  
        }
  
    },
    watch:{
        question(value, oldValue){
            this.isValidQuestion=false
            console.log({value}, 'console.log de componente')

            if( value.includes('?') ){
                this.isValidQuestion=true
                this.getAnswer()
            }

        }
    }

}
</script>


<style scoped>

    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 30px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }

</style>