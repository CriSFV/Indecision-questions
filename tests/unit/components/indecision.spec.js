import { shallowMount } from '@vue/test-utils';
import Indecision from '@/components/Indecision';

describe('Indecion Component', ()=>{
    let wrapper
    let clgSpy
    let getAnswerSpy
    let input

    global.fetch = jest.fn(()=>Promise.resolve({
        json: ()=> Promise.resolve({
            'answer':'yes',
            'forced':false,
            'image':'https://yesno.wtf/assets/yes/2.gif'
        })
    }))
// para hacer test con axios:
//     import axios from 'axios';
 
// jest.mock('axios');
 
// const data = {
//       data: {
//         hits: [
//           {
//             objectID: '1',
//             title: 'a',
//           },
//           {
//             objectID: '2',
//             title: 'b',
//           },
//         ],
//       },
//     };
 
//     axios.get.mockImplementationOnce(() => Promise.resolve(data));

    beforeEach(()=>{
        wrapper= shallowMount(Indecision) 
        clgSpy = jest.spyOn(console, 'log') // console es el objeto y log es el objeto. Esta es una función espía
        getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer') // queremos ver que no sea llamado porque no se ha dado al "?". Con el vm llamamos como al DOM virtual... podemos utilizar tambien vm.$data(es un {} al que está observando Vue), vm.$props, vm.$el
        input = wrapper.find('input')
        jest.clearAllMocks() // esta linea es para que limpien las variables despues de un test
      })

    
    test('Debe hacer match con el snapshot', () => {   
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('El input no debe disparar nada (console.log)', async ()=>{

        // con setValue le damos el valo
       await input.setValue('Hola Mundo')
       // esperamos que el spy sea llamado una vece
       expect(clgSpy).toHaveBeenCalledTimes(1)
       // comprobar que el getAnswer no hay sido disparado
       //expect( getAnswerSpy ).toHaveBeenCalledTimes(0)
       expect( getAnswerSpy ).not.toHaveBeenCalled() // esto es lo mismo que la línea anterior
    })

    test ('disparar fetch al escribir "?"', async ()=>{  
        await input.setValue('?')
        expect(getAnswerSpy).toHaveBeenCalledTimes(1)
    })
    test('pruegas en getAnswer', async()=>{
        await wrapper.vm.getAnswer()
        const img = wrapper.find('img')
        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect(wrapper.vm.answer).toBe('Si!')
        console.log(wrapper.vm.img)
        console.log(wrapper.vm.answer)
    })
    test('pruebas en getAnswer - Fallo en API', async()=>{

        fetch.mockImplementationOnce(()=>Promise.reject('API is down')) // creamos error en api de manera manual
        await wrapper.vm.getAnswer()
        const img = wrapper.find('img')
        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe('No se pudo cargar la respuesta del API')

    })
}) 