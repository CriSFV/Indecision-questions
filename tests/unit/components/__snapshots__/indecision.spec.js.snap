// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecion Component Debe hacer match con el snapshot 1`] = `
<div><img src="" style="display: none;">
  <div class="bg-dark"></div>
  <div class="indecision-container"><input type="text" placeholder="Realiza una pregunta">
    <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
    <!--v-if-->
  </div>
</div>
`;
